tensorflow_quantum/core/ops/math_ops/tfq_inner_product_adj_grad.cc

```diff
@@ -0,0 +1,478 @@
+
+    // track metadata.
+    std::vector<std::vector<tfq::GateMetaData>> gate_meta(
+        programs.size(), std::vector<tfq::GateMetaData>({}));
+
+    // Construct qsim circuits.
+    std::vector<std::vector<std::vector<qsim::GateFused<QsimGate>>>>
+        partial_fused_circuits(
+            programs.size(),
+            std::vector<std::vector<qsim::GateFused<QsimGate>>>({}));
+
+    // track gradients
+    std::vector<std::vector<GradientOfGate>> gradient_gates(
+        programs.size(), std::vector<GradientOfGate>({}));
+
+    auto construct_f = [&](int start, int end) {
+      for (int i = start; i < end; i++) {
+        OP_REQUIRES_OK(
+            context, QsimCircuitFromProgram(programs[i], maps[i], num_qubits[i],
+                                            &qsim_circuits[i],
+                                            &fused_circuits[i], &gate_meta[i]));
+
+        CreateGradientCircuit(qsim_circuits[i], gate_meta[i],
+                              &partial_fused_circuits[i], &gradient_gates[i]);
+      }
+    };
+
+    const int num_cycles = 1000;
+    context->device()->tensorflow_cpu_worker_threads()->workers->ParallelFor(
+        output_dim_batch_size, num_cycles, construct_f);
+
+    // Construct qsim circuits for other_programs.
+    std::vector<std::vector<QsimCircuit>> other_qsim_circuits(
+        output_dim_batch_size,
+        std::vector<QsimCircuit>(output_dim_internal_size, QsimCircuit()));
+    std::vector<std::vector<QsimFusedCircuit>> other_fused_circuits(
+        output_dim_batch_size,
+        std::vector<QsimFusedCircuit>(output_dim_internal_size,
+                                      QsimFusedCircuit({})));
+
+    auto construct_f2 = [&](int start, int end) {
+      for (int i = start; i < end; i++) {
+        int ii = i / output_dim_internal_size;
+        int jj = i % output_dim_internal_size;
+        Status status = QsimCircuitFromProgram(
+            other_programs[ii][jj], {}, num_qubits[ii],
+            &other_qsim_circuits[ii][jj], &other_fused_circuits[ii][jj]);
+        OP_REQUIRES(context, status.ok(),
+                    tensorflow::errors::InvalidArgument(absl::StrCat(
+                        "Found symbols in other_programs.",
+                        "No symbols are allowed in these circuits.")));
+      }
+    };
+
+    context->device()->tensorflow_cpu_worker_threads()->workers->ParallelFor(
+        output_dim_batch_size * output_dim_internal_size, num_cycles,
+        construct_f2);
+
+    int max_num_qubits = 0;
+    for (const int num : num_qubits) {
+      max_num_qubits = std::max(max_num_qubits, num);
+    }
+
+    // Cross reference with standard google cloud compute instances
+    // Memory ~= 2 * num_threads * (2 * 64 * 2 ** num_qubits in circuits)
+    // e2s2 = 2 CPU, 8GB -> Can safely do 25 since Memory = 4GB
+    // e2s4 = 4 CPU, 16GB -> Can safely do 25 since Memory = 8GB
+    // ...
+    if (max_num_qubits >= 26 || output_dim_batch_size == 1) {
```

Comment:
Probably want to lower this to 24 since you are creating 4 state vectors inside.

Reply:
Done.